### PR TITLE
PR: Issue #22 and #16

### PR DESCRIPTION
**Pull request by aloker**
Solves two issues (and more)

- Fully implement Relay pagination specs. Fixes several issues in the code; makes it more testable; makes pagination algorithm usable in data fetching; fixes issue #22; 100% test coverage of all changed and added code (60+ tests) See commit message for details.
- Target frameworks changed to netstandard1.3 and net45 (#16)

Note:  this implementation of the pagination algorithm supports the case of passing in `first` and `last`. However, GraphQL-dotnet does not allow this scenario. 

See:

https://github.com/graphql-dotnet/graphql-dotnet/blob/3e937ec208477affdfa22bde6bde3e54d1ffae6e/src/GraphQL/Builders/ConnectionBuilder.cs#L172-L175

This case is discouraged by Facebook (see https://facebook.github.io/relay/graphql/connections.htm#sec-undefined.PageInfo). Still, this PR correctly implements it. If the GraphQL implementation supports it in the future, this library is prepared.

### CODE REVIEW

File: src/GraphQL.Relay.Test/Utilities/EdgeRangeTests.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        [Theory]
+        [InlineData(10, 9)]
+        [InlineData(10, 8)]
+        public void Ctor_WhenEndOffsetIsLessThanStartOffset_ClampsEndOffsetToOneLessThanStart(int start, int end)
+        {
+            var range = new EdgeRange(start, end);
+            Assert.Equal(start, range.StartOffset);
+            Assert.Equal(start - 1, range.EndOffset);
+            Assert.Equal(0, range.Count);
+            Assert.True(range.IsEmpty);
```
Comment (jquense) [inline src/GraphQL.Relay.Test/Utilities/EdgeRangeTests.cs]: when does this happen? and why is it better to handle it this vs a hard error?

File: src/GraphQL.Relay.Test/Utilities/RelayPaginationTests.cs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            var range = RelayPagination.CalculateEdgeRange(edgeCount, after: ConnectionUtils.OffsetToCursor(afterOffset));
+            Assert.Equal(0, range.StartOffset);
+        }
+
+        [Theory]
+        [InlineData(20, -3)]
+        [InlineData(20, -2)]
+        [InlineData(20, 21)]
+        [InlineData(20, 22)]
+        public void Apply_IfBeforeIsOutOfRange_SilentlyUsesTotalCountAsBoundary(int edgeCount, int afterOffset)
```
Comment (jquense) [inline src/GraphQL.Relay.Test/Utilities/RelayPaginationTests.cs]: might be worth adding some cases for after at the end or start of a range, or length=1 range (unless i've missed those tests)

File: src/GraphQL.Relay.Test/Utilities/EdgeRangeTests.cs (hunk 1 above)
Comment (aloker) [reply src/GraphQL.Relay.Test/Utilities/EdgeRangeTests.cs]: Handling this case gracefully allows the code to deal with the scenario where `before` < `after`. Those cases should in my opinion just return an empty result and not throw an exception. EdgeRange considers all cases where end < start to be empty. I've add another test case that reflects this.

File: src/GraphQL.Relay.Test/Utilities/RelayPaginationTests.cs (hunk 2 above)
Comment (aloker) [reply src/GraphQL.Relay.Test/Utilities/RelayPaginationTests.cs]: I've added a couple more test cases to check for more edge cases (and I indeed found a minor bug in the EdgeRange constructor)

**Comment by aloker** [comment]
NB: I accidentally bumped the assembly version in GraphQL.Relay.csproj.

**Comment by aloker** [comment]
Do you need anything more from me?

**Comment by jquense** [comment]
thanks!
